client/src/NetworkProxy.ts

```diff
@@ -81,37 +146,223 @@ export class NetworkProxy {
 			}, {
 				timeout: 2000
 			});
+
+			if (!proxyAddress) {
+				this.enabledProxyAddress = undefined;
+			} else {
+				this.enabledProxyAddress = proxyAddress;
+			}
 		} catch(e) {
-			await ApplicationRequestProxy.stop();
+			await this.stop();
 		}
 	}
 
+
 	public async stop() {
-		if (!this.proxyAddress) {
+		if (!this.enabledProxyAddress) {
 			this.debugLog('Proxy was not running');
 			// If we weren't running nothing to stop
 			return;
 		}
-		await this.pause();
-		await ApplicationRequestProxy.stop();
+
+		await this.setRokuProxyAddress(null);
+		await new Promise((resolve) => {
+			if (!this.server) {
+				resolve(null);
+				return;
+			}
+
+			this.server.on('close', function() {
+				resolve(null);
+			});
+
+			this.server.close();
+
+			this.savedProxyAddress = undefined;
+		});
+
 		this.debugLog('Proxy stopped');
 	}
 
-	public reloadConfig(configFilePath) {
-		return ApplicationRequestProxy.reloadConfig(configFilePath);
+
+	private debugLog(message: string, ...args) {
+		if (this.getConfig()?.clientDebugLogging) {
+			console.log(`[NetworkProxy] ${message}`, ...args);
+		}
 	}
 
-	public addBreakPointListener(onProxyRequestCallback: OnProxyRequestCallback) {
-		return ApplicationRequestProxy.addBreakPointListener(onProxyRequestCallback);
+
+	/**
+	 * Below method is triggered by the middleware helps us to overwrite the host.
+	 */
+	private rewriteHost(req) {
+		const urlParts = url.parse(req.originalUrl.replace('/;', ''));
+
+		// Used to proxy request to Charles for development purposes
+		const forwardProxy = this.getConfig()?.forwardProxy;
+		if (forwardProxy) {
+			const urlParts = url.parse(forwardProxy);
+			return {
+				host: urlParts.hostname ?? '',
+				protocol: urlParts.protocol ?? '',
+				port: +(urlParts.port ?? 8888)
+			};
+		}
+
+		return {
+			host: urlParts.hostname ?? '',
+			protocol: urlParts.protocol ?? '',
+			port: urlParts.port ?? urlParts.protocol === 'https:' ? 443 : 80
+		};
 	}
 
-	public observeRequest(url: string, onProxyResponseCallback: OnProxyResponseCallback) {
-		return ApplicationRequestProxy.observeRequest(url, onProxyResponseCallback);
+
+	/**
+	 * Below method is triggered by the middleware that helps us to overwrite the path.
+	 */
+	private rewritePath(path, req) {
+		const urlParts = url.parse(req.originalUrl.replace('/;', ''));
```

Comment:
Is this semicolon the standard way of writing proxy URLs? Like is this how it works for charles as well? If not, it might be good to align on whatever "normal" proxy convention is out there (i don't know, haven't used them much).

Reply:
Yes correct that is how it is done in Charles as well.